File: Desktop/optirh/src/main/java/tn/nexus/Entities/Recrutement/Demande.java
package tn.nexus.Entities.Recrutement;

import java.sql.Date;
import java.sql.Timestamp;

public class Demande {

    private int id;
    private int utilisateurId;
    private int offreId;

    public enum Statut {
        ACCEPTEE,
        REFUSEE,
        EN_ATTENTE;
    }
    private Statut statut; // "En attente", "Acceptée", "Refusée"
    private Timestamp date;
    private String description;
    private String fichierPieceJointe;
    private String nomComplet;
    private String email;
    private String telephone;
    private String adresse;
    private Date dateDebutDisponible;
    private String situationActuelle;

    public Demande() {}

    public Demande(int id, int utilisateurId, int offreId, Statut statut, Timestamp date, String description, String fichierPieceJointe, String nomComplet, String email, String telephone, String adresse, Date dateDebutDisponible, String situationActuelle) {
        this.id = id;
        this.utilisateurId = utilisateurId;
        this.offreId = offreId;
        this.statut = statut;
        this.date = date;
        this.description = description;
        this.fichierPieceJointe = fichierPieceJointe;
        this.nomComplet = nomComplet;
        this.email = email;
        this.telephone = telephone;
        this.adresse = adresse;
        this.dateDebutDisponible = dateDebutDisponible;
        this.situationActuelle = situationActuelle;
    }

    public Demande(int utilisateurId, int offreId, Statut statut, Timestamp date, String description, String fichierPieceJointe, String nomComplet, String email, String telephone, String adresse, Date dateDebutDisponible, String situationActuelle) {
        this.utilisateurId = utilisateurId;
        this.offreId = offreId;
        this.statut = statut;
        this.date = date;
        this.description = description;
        this.fichierPieceJointe = fichierPieceJointe;
        this.nomComplet = nomComplet;
        this.email = email;
        this.telephone = telephone;
        this.adresse = adresse;
        this.dateDebutDisponible = dateDebutDisponible;
        this.situationActuelle = situationActuelle;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getUtilisateurId() {
        return utilisateurId;
    }

    public void setUtilisateurId(int utilisateurId) {
        this.utilisateurId = utilisateurId;
    }

    public int getOffreId() {
        return offreId;
    }

    public void setOffreId(int offreId) {
        this.offreId = offreId;
    }

    public Statut getStatut() {
        return statut;
    }

    public void setStatut(Statut statut) {
        this.statut = statut;
    }

    public Timestamp getDate() {
        return date;
    }

    public void setDate(Timestamp date) {
        this.date = date;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getFichierPieceJointe() {
        return fichierPieceJointe;
    }

    public void setFichierPieceJointe(String fichierPieceJointe) {
        this.fichierPieceJointe = fichierPieceJointe;
    }

    public String getNomComplet() {
        return nomComplet;
    }

    public void setNomComplet(String nomComplet) {
        this.nomComplet = nomComplet;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getTelephone() {
        return telephone;
    }

    public void setTelephone(String telephone) {
        this.telephone = telephone;
    }

    public String getAdresse() {
        return adresse;
    }

    public void setAdresse(String adresse) {
        this.adresse = adresse;
    }

    public Date getDateDebutDisponible() {
        return dateDebutDisponible;
    }

    public void setDateDebutDisponible(Date dateDebutDisponible) {
        this.dateDebutDisponible = dateDebutDisponible;
    }

    public String getSituationActuelle() {
        return situationActuelle;
    }

    public void setSituationActuelle(String situationActuelle) {
        this.situationActuelle = situationActuelle;
    }

    @Override
    public String toString() {
        return "Demande{" +
                "id=" + id +
                ", utilisateurId=" + utilisateurId +
                ", offreId=" + offreId +
                ", statut=" + statut +
                ", date=" + date +
                ", description='" + description + '\'' +
                ", fichierPieceJointe='" + fichierPieceJointe + '\'' +
                ", nomComplet='" + nomComplet + '\'' +
                ", email='" + email + '\'' +
                ", telephone='" + telephone + '\'' +
                ", adresse='" + adresse + '\'' +
                ", dateDebutDisponible=" + dateDebutDisponible +
                ", situationActuelle='" + situationActuelle + '\'' +
                '}';
    }
}

File: Web/optiRH/src/Entity/Demande.php
<?php
namespace App\Entity;

use App\Entity\Offre;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity]
#[ORM\Table(name: "demande")]
class Demande
{
    public const STATUT_EN_ATTENTE = 'EN_ATTENTE';
    public const STATUT_ACCEPTEE = 'ACCEPTEE';
    public const STATUT_REFUSEE = 'REFUSEE';
    public const SITUATION_ETUDIANT = 'Etudiant';
    public const SITUATION_EMPLOYE = 'Employé';
    public const SITUATION_AUTRE = 'Autre';

    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(type: "integer")]
    private int $id;

    #[ORM\Column(type: "string", length: 20)]
    #[Assert\Choice(choices: [self::STATUT_EN_ATTENTE, self::STATUT_ACCEPTEE, self::STATUT_REFUSEE],message: "Le statut doit être En Attente, Acceptée ou Refusée.")]
    private string $statut = self::STATUT_EN_ATTENTE;

    #[ORM\Column(type: "datetime", options: ["default" => "CURRENT_TIMESTAMP"])]
    #[Assert\NotBlank(message: "La date est requise.")]
    private \DateTime $date;

    #[ORM\Column(type: "text", nullable: true)]
    #[Assert\Length(
        min: 10,
        minMessage: "La description doit contenir au moins {{ limit }} caractères."
    )]
    private ?string $description = null;

    #[ORM\ManyToOne(targetEntity: Offre::class, inversedBy: "demandes")]
    #[ORM\JoinColumn(nullable: false, onDelete: "CASCADE")]
    private Offre $offre;

    #[ORM\Column(type: "string", length: 255, nullable: true)]
    private ?string $fichierPieceJointe = null;

    #[ORM\Column(type: "string", length: 255)]
    #[Assert\NotBlank(message: "Le nom complet est requis.")]
    #[Assert\Length(max: 255, maxMessage: "Le nom complet ne peut pas dépasser {{ limit }} caractères.")]
    private string $nomComplet;

    #[ORM\Column(type: "string", length: 255)]
    #[Assert\NotBlank(message: "L'email est requis.")]
    #[Assert\Email(message: "L'email doit être valide.")]
    #[Assert\Length(max: 255, maxMessage: "L'email ne peut pas dépasser {{ limit }} caractères.")]
    private string $email;

    #[ORM\Column(type: "string", length: 20)]
    #[Assert\NotBlank(message: "Le téléphone est requis.")]
    #[Assert\Length(max: 8, maxMessage: "Le téléphone doit avoir {{ limit }} caractères.")]
    #[Assert\Regex(
        pattern: "/^[0-9]{8}$/",
        message: "Le téléphone ne doit contenir que 8  chiffres."
    )]
    private string $telephone;

    #[ORM\Column(type: "string", length: 255, nullable: true)]
    #[Assert\Length(max: 255, maxMessage: "L'adresse ne peut pas dépasser {{ limit }} caractères.")]
    private ?string $adresse = null;

    #[ORM\Column(type: "date", nullable: true)]
    #[Assert\GreaterThanOrEqual("today", message: "La date de début disponible doit être aujourd'hui ou dans le futur.")]
    private ?\DateTimeInterface $dateDebutDisponible = null;

    #[ORM\Column(type: "string", length: 100, nullable: true)]
    #[Assert\Choice(
        choices: [self::SITUATION_ETUDIANT, self::SITUATION_EMPLOYE, self::SITUATION_AUTRE],
        message: "La situation actuelle doit être Etudiant, Employé ou Autre."
    )]
    private ?string $situationActuelle = null;

    public function __construct()
    {
        $this->date = new \DateTime();
        $this->statut = 'EN_ATTENTE';
    }

    public function getId(): int
    {
        return $this->id;
    }

    public function setId(int $id): void
    {
        $this->id = $id;
    }

    public function getStatut(): string
    {
        return $this->statut;
    }

    public function setStatut(string $statut): void
    {
        $this->statut = $statut;
    }

    public function getDate(): \DateTime
    {
        return $this->date;
    }

    public function setDate(\DateTime $date): void
    {
        $this->date = $date;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(?string $description): void
    {
        $this->description = $description;
    }

    public function getOffre(): \App\Entity\Offre
    {
        return $this->offre;
    }

    public function setOffre(\App\Entity\Offre $offre): void
    {
        $this->offre = $offre;
    }

    public function getFichierPieceJointe(): ?string
    {
        return $this->fichierPieceJointe;
    }

    public function setFichierPieceJointe(?string $fichierPieceJointe): void
    {
        $this->fichierPieceJointe = $fichierPieceJointe;
    }

    public function getNomComplet(): string
    {
        return $this->nomComplet;
    }

    public function setNomComplet(string $nomComplet): void
    {
        $this->nomComplet = $nomComplet;
    }

    public function getEmail(): string
    {
        return $this->email;
    }

    public function setEmail(string $email): void
    {
        $this->email = $email;
    }

    public function getTelephone(): string
    {
        return $this->telephone;
    }

    public function setTelephone(string $telephone): void
    {
        $this->telephone = $telephone;
    }

    public function getAdresse(): ?string
    {
        return $this->adresse;
    }

    public function setAdresse(?string $adresse): void
    {
        $this->adresse = $adresse;
    }

    public function getDateDebutDisponible(): ?\DateTimeInterface
    {
        return $this->dateDebutDisponible;
    }

    public function setDateDebutDisponible(?\DateTimeInterface $dateDebutDisponible): void
    {
        $this->dateDebutDisponible = $dateDebutDisponible;
    }

    public function getSituationActuelle(): ?string
    {
        return $this->situationActuelle;
    }

    public function setSituationActuelle(?string $situationActuelle): void
    {
        $this->situationActuelle = $situationActuelle;
    }

}
